File: PRNumberBiz/app/src/main/java/com/pplus/prnumberbiz/core/location/LocationUtil.java
package com.pplus.prnumberbiz.core.location;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationManager;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings;
import androidx.core.content.ContextCompat;

import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationListener;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.pple.pplus.utils.part.logs.LogUtil;
import com.pple.pplus.utils.part.pref.PreferenceUtil;
import com.pple.pplus.utils.part.utils.StringUtils;
import com.pplus.prnumberbiz.Const;
import com.pplus.prnumberbiz.R;
import com.pplus.prnumberbiz.apps.common.builder.AlertBuilder;
import com.pplus.prnumberbiz.apps.common.builder.OnAlertResultListener;
import com.pplus.prnumberbiz.apps.common.builder.data.AlertData;
import com.pplus.prnumberbiz.core.network.model.dto.LocationData;

/**
 * Created by 김종경 on 2016-08-26.
 */
public class LocationUtil{

    private static LocationData mSpecifyLocationData;
    private static GoogleApiClient mGoogleApiClient;

    public static double DEFAULT_LATITUDE = 37.4899469;
    public static double DEFAULT_LONGITUDE = 127.0318931;

    public static void init(Context context, GoogleApiClient.ConnectionCallbacks callbacks){

        if(mGoogleApiClient == null) {
            mGoogleApiClient = new GoogleApiClient.Builder(context).addApi(LocationServices.API).addConnectionCallbacks(callbacks).build();
        } else {
            mGoogleApiClient.registerConnectionCallbacks(callbacks);
        }
    }

    public static boolean isConnected(){

        return mGoogleApiClient.isConnected();
    }

    public static void connect(){

        if(!mGoogleApiClient.isConnected()) {
            mGoogleApiClient.connect();
        }
    }

    public static void disConnect(){

        if(mGoogleApiClient.isConnected()) {
            mGoogleApiClient.disconnect();
        }
    }

    public static Location getCurrentLocation(Activity activity){//위치값 가져오기
        // 퍼미션 확인 후 요청
        if(mGoogleApiClient.isConnected()) {
            LogUtil.e("mGoogleApiClient", "isConnected");
            if(ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                Location currentLocation = LocationServices.FusedLocationApi.getLastLocation(mGoogleApiClient);
                return currentLocation;
            }
            return null;
        } else {
            return null;
        }
    }

    public static void startLocationUpdates(final Activity activity, final LocationListener listener){//위치값 요청하기

        // TODO Null point exception이 발생합니다.
        try {
            //            Intent myIntent = new Intent( Settings.ACTION_LOCATION_SOURCE_SETTINGS);
            //            context.startActivity(myIntent);
            if(mGoogleApiClient.isConnected()) {
                if(ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                    LocationRequest locationRequest = new LocationRequest();
                    locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

                    // 위치 정보 요청후 30초간 응답이 없다면 사용자에게 네트워크 정보를 요청 하도록 합니다.
                    final Handler handler = new Handler(Looper.myLooper());

                    final Runnable timeOutRunnable = new Runnable(){

                        @Override
                        public void run(){


                            AlertBuilder.Builder builder = new AlertBuilder.Builder().setAutoCancel(false).setBackgroundClickable(false);
                            builder.addContents(new AlertData.MessageData(activity.getString(R.string.msg_location_error1), AlertBuilder.MESSAGE_TYPE.TEXT, 1));
                            builder.addContents(new AlertData.MessageData(activity.getString(R.string.msg_location_error2), AlertBuilder.MESSAGE_TYPE.TEXT, 2));
                            builder.addContents(new AlertData.MessageData(activity.getString(R.string.msg_location_error3), AlertBuilder.MESSAGE_TYPE.TEXT, 2));
                            builder.setLeftText(activity.getString(R.string.word_cancel)).setRightText(activity.getString(R.string.msg_setting)).setOnAlertResultListener(new OnAlertResultListener(){

                                @Override
                                public void onCancel(){

                                    LocationData locationData = getLastLocation(activity);

                                    Location location = new Location("location");

                                    if(locationData != null) {
                                        location.setLatitude(locationData.getLatitude());
                                        location.setLongitude(locationData.getLongitude());

                                    } else {
                                        location = getDefaultData();
                                    }

                                    listener.onLocationChanged(location);

                                }

                                @Override
                                public void onResult(AlertBuilder.EVENT_ALERT event_alert){

                                    switch (event_alert) {
                                        case RIGHT:

                                            Intent locationIntent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
                                            activity.startActivityForResult(locationIntent, Const.REQ_LOCATION_CODE);

                                            break;
                                    }

                                    LocationData locationData = getLastLocation(activity);

                                    Location location = new Location("location");

                                    if(locationData != null) {
                                        location.setLatitude(locationData.getLatitude());
                                        location.setLongitude(locationData.getLongitude());

                                    } else {
                                        location = getDefaultData();
                                    }

                                    listener.onLocationChanged(location);


                                }
                            }).builder().show(activity);
                        }

                    };

                    handler.postDelayed(timeOutRunnable, 5 * 1000);

                    LocationServices.FusedLocationApi.requestLocationUpdates(mGoogleApiClient, locationRequest, new LocationListener(){

                        @Override
                        public void onLocationChanged(Location location){

                            listener.onLocationChanged(location);
                            handler.removeCallbacks(timeOutRunnable);

//                            disConnect();
//                            stopLocationUpdates();
                        }
                    });
                }
            }

        } catch (NullPointerException e) {

        }

    }

    /**
     * Removes location updates from the FusedLocationApi.
     */
    public static void stopLocationUpdates(){
//        disConnect();
//        if(mGoogleApiClient.isConnected()){
//            disConnect();
//
            LocationServices.FusedLocationApi.removeLocationUpdates(mGoogleApiClient, new LocationListener(){

                @Override
                public void onLocationChanged(Location location){

                }
            });
//        }



    }

    public static boolean isLocationEnabled(Context context){

        return (isGpsLocationEnabled(context) || isNetworkLocationEnabled(context));
    }

    public static boolean isGpsLocationEnabled(Context context){

        LocationManager locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
    }

    public static boolean isNetworkLocationEnabled(Context context){

        LocationManager locationManager = (LocationManager) context.getSystemService(Context.LOCATION_SERVICE);
        return (locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER) || locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER));
    }

    public static void setLastLocation(Context context, double latitude, double longitude){

        PreferenceUtil.getDefaultPreference(context).put(Const.LATITUDE, "" + latitude);
        PreferenceUtil.getDefaultPreference(context).put(Const.LONGITUDE, "" + longitude);
    }

    public static LocationData getLastLocation(Context context){

        LocationData data = new LocationData();
        if(StringUtils.isEmpty(PreferenceUtil.getDefaultPreference(context).getString(Const.LATITUDE))) {
            return null;
        }
        if(StringUtils.isEmpty(PreferenceUtil.getDefaultPreference(context).getString(Const.LONGITUDE))) {
            return null;
        }
        data.setLatitude(Double.parseDouble(PreferenceUtil.getDefaultPreference(context).getString(Const.LATITUDE)));
        data.setLongitude(Double.parseDouble(PreferenceUtil.getDefaultPreference(context).getString(Const.LONGITUDE)));
        return data;
    }

    public static Location getDefaultData() {
        Location location = new Location("location");
        location.setLatitude(DEFAULT_LATITUDE);
        location.setLongitude(DEFAULT_LONGITUDE);

        return location;
    }

    public static LocationData getDefaultLocationData() {
        LocationData locationData = new LocationData();
        locationData.setLatitude(DEFAULT_LATITUDE);
        locationData.setLongitude(DEFAULT_LONGITUDE);
        return locationData;
    }

    public static void setSpecifyLocationData(LocationData specifyLocationData){

        LocationUtil.mSpecifyLocationData = specifyLocationData;
    }

    public static LocationData getSpecifyLocationData(){

        // 값이 없는 경우 Default값 생성
        if(mSpecifyLocationData == null) {
            mSpecifyLocationData = getDefaultLocationData();
        }
        return mSpecifyLocationData;
    }
}

File: LuckyBol/app/src/main/java/com/pplus/luckybol/core/location/LocationUtil.kt
package com.pplus.luckybol.core.location

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.IntentSender.SendIntentException
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.*
import com.pplus.luckybol.Const
import com.pplus.luckybol.R
import com.pplus.luckybol.apps.common.builder.AlertBuilder
import com.pplus.luckybol.apps.common.builder.AlertBuilder.EVENT_ALERT
import com.pplus.luckybol.apps.common.builder.OnAlertResultListener
import com.pplus.luckybol.apps.common.builder.data.AlertData.MessageData
import com.pplus.luckybol.apps.common.ui.base.BaseActivity
import com.pplus.luckybol.core.network.model.dto.LocationData
import com.pplus.utils.part.logs.LogUtil
import com.pplus.utils.part.pref.PreferenceUtil
import com.pplus.utils.part.utils.StringUtils

/**
 * Created by 김종경 on 2016-08-26.
 */
object LocationUtil {
    // 값이 없는 경우 Default값 생성
//        if (mSpecifyLocationData == null) {
//            mSpecifyLocationData = getDefaultLocationData();
//        }
    var specifyLocationData: LocationData? = null

    @JvmField
    var DEFAULT_LATITUDE = 37.4876057

    @JvmField
    var DEFAULT_LONGITUDE = 127.0107591

    @JvmStatic
    fun getCurrentLocation(activity: Activity?): Location? { //위치값 가져오기
        // 퍼미션 확인 후 요청
        return if (ContextCompat.checkSelfPermission(activity!!, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            LocationServices.getFusedLocationProviderClient(activity).lastLocation.result
        } else {
            null
        }
    }

    fun getLastLocation(activity: BaseActivity, listener: LocationListener) {
        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {

            return
        }
        LocationServices.getFusedLocationProviderClient(activity).lastLocation.addOnSuccessListener {
            listener.onLocationChanged(it)
        }.addOnFailureListener {
            startLocationUpdates(activity, listener)
        }
    }

    var mLocationCallback: LocationCallback? = null

    interface LocationResultListener {

        fun result(location:Location?)
    }
    fun startLocationOnly(activity: BaseActivity, listener: LocationResultListener?) { //위치값 요청하기

        // TODO Null point exception이 발생합니다.
        try {
            if (ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {

                // 위치 정보 요청후 30초간 응답이 없다면 사용자에게 네트워크 정보를 요청 하도록 합니다.
                val handler = Handler(Looper.myLooper()!!)
                val timeOutRunnable = Runnable {
//                    val builder = AlertBuilder.Builder().setAutoCancel(false).setBackgroundClickable(false)
//                    builder.addContents(MessageData(activity.getString(R.string.msg_location_error1), AlertBuilder.MESSAGE_TYPE.TEXT, 1))
//                    builder.addContents(MessageData(activity.getString(R.string.msg_location_error2), AlertBuilder.MESSAGE_TYPE.TEXT, 2))
//                    builder.addContents(MessageData(activity.getString(R.string.msg_location_error3), AlertBuilder.MESSAGE_TYPE.TEXT, 2))
//                    builder.setLeftText(activity.getString(R.string.word_cancel)).setRightText(activity.getString(R.string.msg_setting)).setOnAlertResultListener(object : OnAlertResultListener {
//                        override fun onCancel() {
//                            listener?.result(null)
//                        }
//
//                        override fun onResult(event_alert: EVENT_ALERT) {
//
//                            if(mLocationCallback != null){
//                                LocationServices.getFusedLocationProviderClient(activity).removeLocationUpdates(mLocationCallback)
//                            }
//
//                            when (event_alert) {
//                                EVENT_ALERT.RIGHT -> {
//                                    val locationIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
//                                    activity.locationLauncher.launch(locationIntent)
//                                }
//                                EVENT_ALERT.LEFT -> {
//                                    listener?.result(null)
//                                }
//                            }
//                        }
//                    }).builder().show(activity)
                    listener?.result(null)
                }

                val locationRequest = LocationRequest.create()
                locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
                locationRequest.numUpdates = 1
                locationRequest.interval = 10000
                locationRequest.fastestInterval = 5000
                handler.postDelayed(timeOutRunnable, 10000)
                val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
                val settingsClient = LocationServices.getSettingsClient(activity)
                val task = settingsClient.checkLocationSettings(builder.build())

                mLocationCallback = object : LocationCallback() {
                    private var isLocationUpdates = false
                    override fun onLocationResult(locationResult: LocationResult) {
                        isLocationUpdates = true
                        if(mLocationCallback != null){
                            LocationServices.getFusedLocationProviderClient(activity).removeLocationUpdates(mLocationCallback!!)
                        }
                        val lastLocation = locationResult.lastLocation
                        LogUtil.e("onLocationResult", "onLocationResult")
                        val h = Handler()
                        h.post {
                            handler.removeCallbacks(timeOutRunnable)
                            listener?.result(lastLocation)
                        }
                    }

                    override fun onLocationAvailability(locationAvailability: LocationAvailability) {
                        if (locationAvailability.isLocationAvailable) {
                            LogUtil.e("onLocationAvailability", "locationAvailability is true")
                            try {
                            } catch (e: Exception) {
                                LogUtil.e("onLocationAvailability", e.toString())
                            }
                        }
                    }
                }

                task.addOnSuccessListener(activity) { locationSettingsResponse ->
                    LogUtil.e("OnSuccessListener", "onSuccess {}", locationSettingsResponse.locationSettingsStates?.isLocationPresent)
                    LocationServices.getFusedLocationProviderClient(activity).requestLocationUpdates(locationRequest, mLocationCallback!!, Looper.myLooper()!!)
                }.addOnFailureListener(activity) { e ->
                    LogUtil.e("OnFailureListener", "onFailure {}", e.toString())
                    val statusCode = (e as ApiException).statusCode
                    when (statusCode) {
                        LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {
                            handler.removeCallbacks(timeOutRunnable)
                            try {
                                //위치 설정이 만족스럽지 않을 때 감지.
                                // 그러나 사용자에게 대화상자를  보여줌으로써 고칠 수있습니다.
                                val rae = e as ResolvableApiException
                                //startResolutionForResult ()를 호출하여 대화 상자를 표시합니다.
                                rae.startResolutionForResult(
                                    activity,
                                    Const.REQ_LOCATION_CODE)
                            } catch (sie: SendIntentException) {
                                LogUtil.e("debug", "PendingIntent unable to execute request.")
                            }
                        }
                        LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> LogUtil.e("SETTINGS_CHANGE_UNAVAILABLE", "SETTINGS_CHANGE_UNAVAILABLE")
                    }
                }
            }
        } catch (e: NullPointerException) {
            LogUtil.e("LocationUtil", e.toString())
        }
    }

    fun startLocationUpdates(activity: BaseActivity, listener: LocationListener?) { //위치값 요청하기

        // TODO Null point exception이 발생합니다.
        try {
            if (ContextCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {

                // 위치 정보 요청후 30초간 응답이 없다면 사용자에게 네트워크 정보를 요청 하도록 합니다.
                val handler = Handler(Looper.myLooper()!!)
                val timeOutRunnable = Runnable {
                    val builder = AlertBuilder.Builder().setAutoCancel(false).setBackgroundClickable(false)
                    builder.addContents(MessageData(activity.getString(R.string.msg_location_error1), AlertBuilder.MESSAGE_TYPE.TEXT, 1))
                    builder.addContents(MessageData(activity.getString(R.string.msg_location_error2), AlertBuilder.MESSAGE_TYPE.TEXT, 2))
                    builder.addContents(MessageData(activity.getString(R.string.msg_location_error3), AlertBuilder.MESSAGE_TYPE.TEXT, 2))
                    builder.setLeftText(activity.getString(R.string.word_cancel)).setRightText(activity.getString(R.string.msg_setting)).setOnAlertResultListener(object : OnAlertResultListener {
                        override fun onCancel() {
                            val locationData = getLastLocation(activity)
                            var location = Location("location")
                            if (locationData != null) {
                                location.latitude = locationData.latitude
                                location.longitude = locationData.longitude
                            } else {
                                location = defaultData
                            }
                            listener!!.onLocationChanged(location)
                        }

                        override fun onResult(event_alert: EVENT_ALERT) {

                            if(mLocationCallback != null){
                                LocationServices.getFusedLocationProviderClient(activity).removeLocationUpdates(mLocationCallback!!)
                            }

                            when (event_alert) {
                                EVENT_ALERT.RIGHT -> {
                                    val locationIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
                                    activity.locationLauncher.launch(locationIntent)
                                }
                                EVENT_ALERT.LEFT -> {
                                    val locationData = getLastLocation(activity)
                                    var location = Location("location")
                                    if (locationData != null) {
                                        location.latitude = locationData.latitude
                                        location.longitude = locationData.longitude
                                    } else {
                                        location = defaultData
                                    }
                                    listener!!.onLocationChanged(location)
                                }
                                else -> {}
                            }
                        }
                    }).builder().show(activity)
                }
                val locationRequest = LocationRequest.create()
                locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
                locationRequest.numUpdates = 1
                locationRequest.interval = 10000
                locationRequest.fastestInterval = 5000
                handler.postDelayed(timeOutRunnable, 10000)
                val builder = LocationSettingsRequest.Builder().addLocationRequest(locationRequest)
                val settingsClient = LocationServices.getSettingsClient(activity)
                val task = settingsClient.checkLocationSettings(builder.build())

                mLocationCallback = object : LocationCallback() {
                    private var isLocationUpdates = false
                    override fun onLocationResult(locationResult: LocationResult) {
                        isLocationUpdates = true
                        if(mLocationCallback != null){
                            LocationServices.getFusedLocationProviderClient(activity).removeLocationUpdates(mLocationCallback!!)
                        }
                        val lastLocation = locationResult.lastLocation
                        LogUtil.e("onLocationResult", "onLocationResult")
                        val h = Handler()
                        h.post {
                            handler.removeCallbacks(timeOutRunnable)
                            listener?.onLocationChanged(lastLocation)
                        }
                    }

                    override fun onLocationAvailability(locationAvailability: LocationAvailability) {
                        if (locationAvailability.isLocationAvailable) {
                            LogUtil.e("onLocationAvailability", "locationAvailability is true")
                            try {
                            } catch (e: Exception) {
                                LogUtil.e("onLocationAvailability", e.toString())
                            }
                        }
                    }
                }

                task.addOnSuccessListener(activity) { locationSettingsResponse ->
                    LogUtil.e("OnSuccessListener", "onSuccess {}", locationSettingsResponse.locationSettingsStates?.isLocationPresent)
                    LocationServices.getFusedLocationProviderClient(activity).requestLocationUpdates(locationRequest, mLocationCallback!!, Looper.myLooper()!!)
                }.addOnFailureListener(activity) { e ->
                    LogUtil.e("OnFailureListener", "onFailure {}", e.toString())
                    val statusCode = (e as ApiException).statusCode
                    when (statusCode) {
                        LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> {
                            handler.removeCallbacks(timeOutRunnable)
                            try {
                                //위치 설정이 만족스럽지 않을 때 감지.
                                // 그러나 사용자에게 대화상자를  보여줌으로써 고칠 수있습니다.
                                val rae = e as ResolvableApiException
                                //startResolutionForResult ()를 호출하여 대화 상자를 표시합니다.
                                rae.startResolutionForResult(
                                        activity,
                                        Const.REQ_LOCATION_CODE)
                            } catch (sie: SendIntentException) {
                                LogUtil.e("debug", "PendingIntent unable to execute request.")
                            }
                        }
                        LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> LogUtil.e("SETTINGS_CHANGE_UNAVAILABLE", "SETTINGS_CHANGE_UNAVAILABLE")
                    }
                }
            }
        } catch (e: NullPointerException) {
            LogUtil.e("LocationUtil", e.toString())
        }
    }

    @JvmStatic
    fun isLocationEnabled(context: Context): Boolean {
        return isGpsLocationEnabled(context) || isNetworkLocationEnabled(context)
    }

    fun isGpsLocationEnabled(context: Context): Boolean {
        val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
    }

    fun isNetworkLocationEnabled(context: Context): Boolean {
        val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        return locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER) || locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
    }

    @JvmStatic
    fun setLastLocation(context: Context?, latitude: Double, longitude: Double) {
        PreferenceUtil.getDefaultPreference(context).put(Const.LATITUDE, "" + latitude)
        PreferenceUtil.getDefaultPreference(context).put(Const.LONGITUDE, "" + longitude)
    }

    @JvmStatic
    fun getLastLocation(context: Context?): LocationData? {
        val data = LocationData()
        if (StringUtils.isEmpty(PreferenceUtil.getDefaultPreference(context).getString(Const.LATITUDE))) {
            return null
        }
        if (StringUtils.isEmpty(PreferenceUtil.getDefaultPreference(context).getString(Const.LONGITUDE))) {
            return null
        }
        data.latitude = PreferenceUtil.getDefaultPreference(context).getString(Const.LATITUDE).toDouble()
        data.longitude = PreferenceUtil.getDefaultPreference(context).getString(Const.LONGITUDE).toDouble()
        return data
    }

    val defaultData: Location
        get() {
            val location = Location("location")
            location.latitude = DEFAULT_LATITUDE
            location.longitude = DEFAULT_LONGITUDE
            return location
        }
    val defaultLocationData: LocationData
        get() {
            val locationData = LocationData()
            locationData.latitude = DEFAULT_LATITUDE
            locationData.longitude = DEFAULT_LONGITUDE
            return locationData
        }
}